### ISSUE TITLE
docker compiled from sources, Illegal instruction on Raspbian (v20.10.5)

### ISSUE THREAD

**Issue by gabrielitos87** [moby/moby#42212]
**Description**
I compiled Docker binaries on my Raspberry Pi 3 from sources using the standard procedure. However, this resulted in binaries that are providing "Illegal instruction" when run.

**Steps to reproduce the issue:**
1. Take a Raspberry Pi 3 with Raspbian OS Buster fully updated.
2. Installed Docker from helping script get-docker.sh
2. git clone https://github.com/docker/docker.git
2. cd docker
3. make build
4. make binary
5. cd bundles/binary-daemon/
6. ./dockerd -v

**Describe the results you received:**
```
root@raspy:/home/gabrielitos/docker/bundles/binary-daemon# ./dockerd -v
Illegal instruction
root@raspy:/home/gabrielitos/docker/bundles/binary-daemon# strace ./dockerd -v
execve("./dockerd", ["./dockerd", "-v"], 0x7ed217a4 /* 16 vars */) = 0
syscall_0x101cc(0x3fb0d6, 0x811ad021, 0x1, 0x18f51b0, 0x101c4, 0x18c9ed5) = -1 ENOSYS (Function not implemented)
--- SIGILL {si_signo=SIGILL, si_code=ILL_ILLOPC, si_addr=0x18f51b4} ---
+++ killed by SIGILL +++
Illegal instruction
```

**Describe the results you expected:**
```
root@raspy:/home/gabrielitos/docker/bundles/binary-daemon# ./dockerd -v
Docker version 20.10.5, build xxxxxx
```

**Additional information you deem important (e.g. issue happens only occasionally):**
When I download the binaries from [https://download.docker.com/linux/static/stable/armhf/docker-20.10.5.tgz](https://download.docker.com/linux/static/stable/armhf/docker-20.10.5.tgz) I obtain the same result. I reported that here: https://github.com/docker/for-linux/issues/1226

**Output of `docker version`:**

```
root@raspy:/home/gabrielitos/docker/bundles/binary-daemon# docker version
Client: Docker Engine - Community
 Version:           20.10.5
 API version:       1.41
 Go version:        go1.13.15
 Git commit:        55c4c88
 Built:             Tue Mar  2 20:18:46 2021
 OS/Arch:           linux/arm
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.5
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       363e9a8
  Built:            Tue Mar  2 20:16:18 2021
  OS/Arch:          linux/arm
  Experimental:     false
 containerd:
  Version:          1.4.4
  GitCommit:        05f951a3781f4f2c1911b05e61c160e9c30eaa8e
 runc:
  Version:          1.0.0-rc93
  GitCommit:        12644e614e25b05da6fd08a38ffa0cfe1903fdec
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0

```

**Output of `docker info`:**

```
root@raspy:/home/gabrielitos/docker/bundles/binary-daemon# docker info   
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Build with BuildKit (Docker Inc., v0.5.1-docker)

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 1
 Server Version: 20.10.5
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 05f951a3781f4f2c1911b05e61c160e9c30eaa8e
 runc version: 12644e614e25b05da6fd08a38ffa0cfe1903fdec
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 5.10.17-v7+
 Operating System: Raspbian GNU/Linux 10 (buster)
 OSType: linux
 Architecture: armv7l
 CPUs: 4
 Total Memory: 924.2MiB
 Name: raspy
 ID: 6GBY:6NYP:C3EG:UGUA:OXSK:EFJY:52KQ:ADR5:LESP:OE7J:TLSG:RIA2
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No memory limit support
WARNING: No swap limit support
WARNING: No kernel memory TCP limit support
WARNING: No oom kill disable support
WARNING: No blkio throttle.read_bps_device support
WARNING: No blkio throttle.write_bps_device support
WARNING: No blkio throttle.read_iops_device support
WARNING: No blkio throttle.write_iops_device support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
```
root@raspy:/home/gabrielitos/docker/bundles/binary-daemon# uname -a
Linux raspy 5.10.17-v7+ #1403 SMP Mon Feb 22 11:29:51 GMT 2021 armv7l GNU/Linux
```

**Comment by gabrielitos87** [comment]
Can it be related to this?
https://stackoverflow.com/questions/55465118/gcc-8-cross-compiler-outputs-armv7-executable-instead-of-armv6

I know that Raspberry Pi 3 is armv7, but maybe Raspbian is compiled for armv6.

**Comment by gabrielitos87** [comment]
I managed to compile the moby/moby build and binary by reverting the Dockerfile to Stretch (and few other changes): see below!

Why Buster does not work for Raspberry Pi 3, whereas Stretch works?? Also the official Docker images available from https://github.com/docker/for-linux are not working...

```
diff --git a/Dockerfile b/Dockerfile
index f5ec77836b..4ac9fb7ebd 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -8,7 +8,7 @@ ARG DEBIAN_FRONTEND=noninteractive
 ARG VPNKIT_VERSION=0.5.0
 ARG DOCKER_BUILDTAGS="apparmor seccomp"
 
-ARG BASE_DEBIAN_DISTRO="buster"
+ARG BASE_DEBIAN_DISTRO="stretch"
 ARG GOLANG_IMAGE="golang:${GO_VERSION}-${BASE_DEBIAN_DISTRO}"
 
 FROM ${GOLANG_IMAGE} AS base
@@ -23,6 +23,7 @@ ARG DEBIAN_FRONTEND
 # Install dependency packages specific to criu
 RUN --mount=type=cache,sharing=locked,id=moby-criu-aptlib,target=/var/lib/apt \
     --mount=type=cache,sharing=locked,id=moby-criu-aptcache,target=/var/cache/apt \
+        rm -rf /var/lib/apt/lists/ && \
         apt-get update && apt-get install -y --no-install-recommends \
             libcap-dev \
             libnet-dev \
@@ -85,6 +86,7 @@ FROM debian:${BASE_DEBIAN_DISTRO} AS frozen-images
 ARG DEBIAN_FRONTEND
 RUN --mount=type=cache,sharing=locked,id=moby-frozen-images-aptlib,target=/var/lib/apt \
     --mount=type=cache,sharing=locked,id=moby-frozen-images-aptcache,target=/var/cache/apt \
+       rm -rf /var/lib/apt/lists/ && \
        apt-get update && apt-get install -y --no-install-recommends \
            ca-certificates \
            curl \
@@ -110,6 +112,7 @@ RUN dpkg --add-architecture armel
 RUN dpkg --add-architecture armhf
 RUN --mount=type=cache,sharing=locked,id=moby-cross-true-aptlib,target=/var/lib/apt \
     --mount=type=cache,sharing=locked,id=moby-cross-true-aptcache,target=/var/cache/apt \
+        rm -rf /var/lib/apt/lists/ && \
         apt-get update && apt-get install -y --no-install-recommends \
             crossbuild-essential-arm64 \
             crossbuild-essential-armel \
@@ -122,6 +125,7 @@ ARG DEBIAN_FRONTEND
 RUN echo 'deb http://deb.debian.org/debian buster-backports main' > /etc/apt/sources.list.d/backports.list
 RUN --mount=type=cache,sharing=locked,id=moby-cross-false-aptlib,target=/var/lib/apt \
     --mount=type=cache,sharing=locked,id=moby-cross-false-aptcache,target=/var/cache/apt \
+        rm -rf /var/lib/apt/lists/ && \
         apt-get update && apt-get install -y --no-install-recommends \
             binutils-mingw-w64 \
             g++-mingw-w64-x86-64 \
@@ -141,6 +145,7 @@ ARG DEBIAN_FRONTEND
 RUN echo 'deb http://deb.debian.org/debian buster-backports main' > /etc/apt/sources.list.d/backports.list
 RUN --mount=type=cache,sharing=locked,id=moby-cross-true-aptlib,target=/var/lib/apt \
     --mount=type=cache,sharing=locked,id=moby-cross-true-aptcache,target=/var/cache/apt \
+        rm -rf /var/lib/apt/lists/ && \
         apt-get update && apt-get install -y --no-install-recommends \
             libapparmor-dev:arm64 \
             libapparmor-dev:armel \
@@ -167,8 +172,10 @@ RUN --mount=type=cache,target=/root/.cache/go-build \
 
 FROM dev-base AS containerd
 ARG DEBIAN_FRONTEND
+RUN echo 'deb http://deb.debian.org/debian stretch-backports main' > /etc/apt/sources.list.d/backports.list
 RUN --mount=type=cache,sharing=locked,id=moby-containerd-aptlib,target=/var/lib/apt \
     --mount=type=cache,sharing=locked,id=moby-containerd-aptcache,target=/var/cache/apt \
+        rm -rf /var/lib/apt/lists/ && \
         apt-get update && apt-get install -y --no-install-recommends \
             libbtrfs-dev
 ARG CONTAINERD_COMMIT
@@ -226,6 +233,7 @@ ARG DEBIAN_FRONTEND
 ARG TINI_COMMIT
 RUN --mount=type=cache,sharing=locked,id=moby-tini-aptlib,target=/var/lib/apt \
     --mount=type=cache,sharing=locked,id=moby-tini-aptcache,target=/var/cache/apt \
+        rm -rf /var/lib/apt/lists/ && \
         apt-get update && apt-get install -y --no-install-recommends \
             cmake \
             vim-common
@@ -268,6 +276,7 @@ RUN ldconfig
 # Do you really need to add another package here? Can it be done in a different build stage?
 RUN --mount=type=cache,sharing=locked,id=moby-dev-aptlib,target=/var/lib/apt \
     --mount=type=cache,sharing=locked,id=moby-dev-aptcache,target=/var/cache/apt \
+        rm -rf /var/lib/apt/lists/ && \
         apt-get update && apt-get install -y --no-install-recommends \
             apparmor \
             aufs-tools \
@@ -330,6 +339,7 @@ ENTRYPOINT ["hack/dind"]
 FROM dev-systemd-false AS dev-systemd-true
 RUN --mount=type=cache,sharing=locked,id=moby-dev-aptlib,target=/var/lib/apt \
     --mount=type=cache,sharing=locked,id=moby-dev-aptcache,target=/var/cache/apt \
+        rm -rf /var/lib/apt/lists/ && \
         apt-get update && apt-get install -y --no-install-recommends \
             dbus \
             dbus-user-session \
```

**Comment by grooverdan** [comment]
Note this appears in release binaries too:
https://doi-janky.infosiftr.net/job/multiarch/job/arm32v7/job/docker/8854/execution/node/55/log/

+ url=https://download.docker.com/linux/static/stable/armhf/docker-20.10.7.tgz
+ wget -O docker.tgz https://download.docker.com/linux/static/stable/armhf/docker-20.10.7.tgz
[0m[91mConnecting to download.docker.com (65.8.228.23:443)
[0m[91msaving to 'docker.tgz'
[0m[91mdocker.tgz            58% |******************              | 35.8M  0:00:00 ETA
[0m[91mdocker.tgz           100% |********************************| 61.5M  0:00:00 ETA
'docker.tgz' saved
[0m[91m+ tar --extract --file docker.tgz --strip-components 1 --directory /usr/local/bin/
[0m[91m+ rm docker.tgz
[0m[91m+ dockerd --version
[0m[91mIllegal instruction (core dumped)

And https://doi-janky.infosiftr.net/job/multiarch/job/arm32v6/job/docker/9422/execution/node/55/log/

 wget -O docker.tgz https://download.docker.com/linux/static/stable/armel/docker-20.10.7.tgz
[0m[91mConnecting to download.docker.com (65.8.228.23:443)
[0m[91msaving to 'docker.tgz'
[0m[91mdocker.tgz             3% |                                | 1919k  0:00:31 ETA
[0m[91mdocker.tgz            53% |*****************               | 33.2M  0:00:01 ETA
[0m[91mdocker.tgz           100% |********************************| 61.6M  0:00:00 ETA
'docker.tgz' saved
[0m[91m+ tar --extract --file docker.tgz --strip-components 1 --directory /usr/local/bin/
[0m[91m+ rm docker.tgz
[0m[91m+ dockerd --version
[0m[91mIllegal instruction (core dumped)

**Comment by alerei** [comment]
I've just spent quite some time investigating this issue and would like to share my insights as I am currently stuck on how and where to continue my research.

### TL;DR: This issue is somehow related to statically linking against glibc. At some point the CPU tries to execute code in the wrong instruction set. I do not have a definitive solution at this point, but some thoughts

1. Dynamically linking cross-compiled binaries (something along the lines of a "dyncross" target) seems to work
2. Reverting to stretch as base image seems to work, as pointed out by @gabrielitos87 (and confirmed by me)
3. I've noticed differences in Debian's `libc6-dev:armhf` and `libc6-dev-armhf-cross` packages, which may be relevant - I would have expected them to be identical
4. Maybe a similar issue would occur for some minimal code example to narrow down possible causes

### Preparation
First of all, here is what I did:

1. Use git tag v20.10.17 to cross-build dockerd for linux/arm/v7
5. Test on Raspberry Pi 
    1. With official Raspbian 11 / armhf
    2. With Debian for Raspberry Pi (2022.01.21 from here https://raspi.debian.net/tested-images/)
6.  Upgrade toolchain to Debian 11 / bullseye (Thanks @gabrielitos87 for sharing your steps - made it quite easy for me)
7. Test on Raspberry Pi
    1. With official Raspbian 11 / armhf
    2. With Debian for Raspberry Pi (2022.01.21 from here https://raspi.debian.net/tested-images/)

All tests lead to a crash on startup with SIGILL. strace shows a nonsensical syscall before the crash:
```
syscall_0x101cc(0x3fb0d6, 0x4109a028, 0x1, 0x1757690, 0x101c4, 0x172c075) = -1 ENOSYS (Function not implemented)
--- SIGILL {si_signo=SIGILL, si_code=ILL_ILLOPC, si_addr=0x17576a0} ---
```

**As a first conclusion, this issue is not related to some Raspbian-specific property, but persists when using official Debian, for which the binary is built.**

### Approaching the issue with gdb
Next, I did the obvious and ran the binary in gdb, which gave me this callstack:

```
#0  0x017576a0 in memchr_ifunc ()
#1  0x0172bc08 in __libc_start_main ()
#2  0x00010384 in _start ()
```
According to `info proc mappings` all these symbols come from the statically linked binary itself, no other dynamic library is loaded until that point. That made me wonder if linking libc dynamically would make things better, so there I went

1. Set ```export IAMSTATIC='false' export LDFLAGS_STATIC_DOCKER=''``` during build
2. Test on Raspberry Pi

and got a working binary!

**Next conclusion: The reason for sure is some issue with linking libc statically**

### Deep-dive with gdb
Let's take a closer look at the last instructions in `__libc_start_main` before the crash 
```
   0x0172bbfe <+150>:	ldr.w	r3, [r8]
   0x0172bc02 <+154>:	adds	r4, #8
   0x0172bc04 <+156>:	ldr	r0, [r6, #0]
   0x0172bc06 <+158>:	blx	r3
=> 0x0172bc08 <+160>:	str.w	r0, [r8]
```
We're looking at an indirect function call here. Some value from some address is loaded from `[r8]` into `r3` and is called as a function with `blx r3`. `blx` not only jumps to that address, but also switches between Thumb2 and ARM instruction set, depending on the target function's address. Even addresses mean ARM code, odd addresses mean Thumb2.

`r3` contains `0x1757690`, which is the address of `memchr_ifunc` (the one that crashes!), and an even address means that instruction set is switched to ARM. `cpsr` register (https://www.keil.com/pack/doc/CMSIS/Core_A/html/group__CMSIS__CPSR.html) confirms this: it is `0x30` before the call and `0x10` after -- Thumb2 is switched to ARM32.

Now, `memchr_ifunc` is really interesting:
```
   0x01757690 <+0>:	tst.w	r0, #4096	; 0x1000
   0x01757694 <+4>:	movw	r3, #32049	; 0x7d31
   0x01757698 <+8>:	movt	r3, #373	; 0x175
   0x0175769c <+12>:	movw	r0, #32289	; 0x7e21
=> 0x017576a0 <+16>:	movt	r0, #373	; 0x175
   0x017576a4 <+20>:	it	ne
   0x017576a6 <+22>:	movne	r0, r3
   0x017576a8 <+24>:	bx	lr
```
While this looks like completely valid code, it is only valid in Thumb2 state - however, we left Thumb2 state due to `memchr_ifunc`'s even address! Instruction-stepping confirms that all the `mov` instructions do anything but modify the content of `r3` and `r0`, because they are executed as ARM code. (What this function is supposed to do is call one of multiple `memchr` implementations depending on the CPU's capabilities, namely `__memchr_neon` and `__memchr_noneon`).

For completeness, let's look at the ARM-disassembly of `memchr_ifunc`:
```
   0x01757690 <+0>:	svcpl	0x0080f410
   0x01757694 <+4>:	teqpl	r1, #74448896	; 0x4700000	; <UNPREDICTABLE>
   0x01757698 <+8>:	cmnne	r5, #192, 4	; <UNPREDICTABLE>
   0x0175769c <+12>:	eorvs	pc, r1, r7, asr #12
=> 0x017576a0 <+16>:	rsbsne	pc, r5, r0, asr #5
   0x017576a4 <+20>:	sadd16mi	r11, r8, r8
   0x017576a8 <+24>:	svclt	0x00004770
```
I have no idea what this is, and neither does my poor Raspberry Pi, which is why we hard-exit with SIGILL at this point.

### Jerry-rigging the binary with gdb
Let's confirm that this would run just fine in Thumb2 state. Basically, what we do is break when the function's address is loaded into `r3` and modify its address by adding 1 beforehand (e.g. `0x1757690` -> `0x1757691`).
```
(gdb) b *0x0172bbfe
(gdb) r
Breakpoint 1, 0x0172bc06 in __libc_start_main ()
(gdb) p/x *$r8
$1 = 0x1757690
(gdb) set *$r8=0x1757691
(gdb) c
Breakpoint 1, 0x0172bc06 in __libc_start_main ()
(gdb) p/x *$r8
$3 = 0x1743114
(gdb) set *$r8=0x1743115
(gdb) c
...
INFO[2022-08-23T11:01:00.804926211Z] Starting up 
...
```
Looks good! The binary is running fine after this.

**Final conclusion: Thumb2 code is executed in ARM state, which leads to the crash**

### Further speculations
I had a look at glibc-2.31's source code (https://sourceware.org/git/?p=glibc.git;a=commit;h=9ea3686266dca3f004ba874745a4087a89682617) and suppose we are crashing in the `ARCH_SETUP_IREL` macro. From a quick look I assume this is expanded to `apply_irel()`, which invokes `elf_irel()` defined in `sysdeps/arm/dl-irel.h`. However, I am not sure about this.

### Conclusion
So here I am, stuck and wondering if I am missing anything completely obvious because I have no explanation how this behavior comes to be. Any ideas, comments, suggestions would be highly appreciated!

**Comment by thaJeztah** [comment]
Let me close this ticket for now, as it looks like it went stale.
